finetuner/tuner/base.py

```diff
@@ -1,112 +1,151 @@
 import abc
-import warnings
-from typing import (
-    Generator,
-    Optional,
-    Union,
-    Tuple,
-    List,
-    Dict,
-)
-
-from ..helper import AnyDNN, AnyTensor, AnyDataLoader, AnyOptimizer, DocumentArrayLike
+from typing import Callable, Generic, List, Optional, Tuple, TypeVar, Union
+
+from .dataset import ClassDataset, SessionDataset
+from .dataset.samplers import RandomClassBatchSampler, SessionBatchSampler
 from .summary import Summary
+from ..helper import AnyDataLoader, AnyDNN, AnyOptimizer, AnyTensor, DocumentSequence
+
+
+LabelType = TypeVar('LabelType')
+
+
+class BaseMiner(abc.ABC, Generic[AnyTensor, LabelType]):
+    @abc.abstractmethod
+    def mine(self, labels: LabelType, distances: AnyTensor) -> Tuple[AnyTensor, ...]:
+        """Generate mined tuples from labels and item distances.
+
+        :param labels: labels of each item
+        :param distances: A tensor matrix of pairwise distance between each two item
+            embeddings
+
+        :return: A tuple of 1D tensors, denoting indices or properties of returned
+            tuples
+        """
+
+
+class BaseClassMiner(BaseMiner[AnyTensor, AnyTensor], Generic[AnyTensor]):
+    @abc.abstractmethod
+    def mine(self, labels: AnyTensor, distances: AnyTensor) -> Tuple[AnyTensor, ...]:
```

Comment:
are we sure this is the wanted interface? isn't embeddings + dist easier to grasp?